core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -183,4 +220,41 @@ public BlockReader createFileReader(String fileId, long offset, boolean position
   @Override
   public void cleanupSession(long sessionId) {
   }
+
+  private class BlockMasterSync implements HeartbeatExecutor {
+    @Override
+    public void heartbeat() throws InterruptedException {
+      final Command cmdFromMaster;
+      BlockMasterClient masterClient = mBlockMasterClientPool.acquire();
+      try {
+        cmdFromMaster = masterClient.heartbeat(mWorkerId.get(),
+            ImmutableMap.of(Constants.MEDIUM_MEM, (long) Constants.GB),
+            ImmutableMap.of(Constants.MEDIUM_MEM, 0L),
+            ImmutableList.of(),
+            ImmutableMap.of(),
+            ImmutableMap.of(),
+            ImmutableList.of());
+      } catch (IOException e) {
+        LOG.warn("failed to heartbeat to master", e);
+        return;
+      } finally {
+        mBlockMasterClientPool.release(masterClient);
+      }
```

Comment:
This piece of code is confusing to new developers.
If exception happens, it will go into the `catch` section, which contains a `return`. Will it `return` immediately without going to the `finally` section?
This is confusing and not easy to understand.

Reply:
No, `finally` will get executed immediately after the return statement, but before it actually returns to the caller.

I agree this piece of code is less ideal here. The acquire and release of a `BlockMasterClient` should be wrapped in a try-with-resource block.

Reply:
I created a `PooledResource` abstraction for this pattern to hopefully make things more clear